### ISSUE TITLE
Feature: Extract to variable/parameter

### ISSUE THREAD

**Issue by StephenWeatherford** [microsoft/vscode-azurearmtools#515]

**Comment by nilshedstrom** [comment]
This seems interesting. Did you start working on this?
Do you mean something like
"Extracting the current selected value to a variable/parameter and replace the current selection with a reference to that variable/parameter"?
Or has it something to do with parameter files too?

**Comment by neilpeterson** [comment]
I do not believe that we have started work on this, but I'm super keen on the capability.

I think that this is definitely part of it.

_"Extracting the current selected value to a variable/parameter and replace the current selection with a reference to that variable/parameter"_

In the case where there is an associated parameter file (upcoming release) where extracting to parameter, we would want to consider also adding the extracted parameter to the parameter file. Ill let @StephenWeatherford weigh in on this.

**Comment by nilshedstrom** [comment]
If a user would extract a parameter from a sku->name of a storage account it would be pretty neat if the allowed values from that field would come up as allowedValues for the new parameter. Or at least give the user the possibility to do so.
![image](https://user-images.githubusercontent.com/785949/79319820-1ca17d00-7f09-11ea-9799-bbe801fc83fd.png)

Is it possible to get the allowed values for a field (such as sku->name for a storage account) from typescript?

**Comment by neilpeterson** [comment]
Yes, this would be super cool. I will let @StephenWeatherford comment on the values question.

One thing we need to consider is validation. I do not think we are resolving expressions at design time. Take the following example, I have a valid storage kind stored in the parameter. I was expecting a warning on line 19, but do not see one.

![image](https://user-images.githubusercontent.com/7844635/79365061-fd9ded80-7efe-11ea-9617-fda1621d46e7.png)

So initially, things look good. However if I update the parameter value to something non-valid, I also do not see a warning. 

![image](https://user-images.githubusercontent.com/7844635/79364552-5b7e0580-7efe-11ea-9998-99488fbb2228.png)

Interesting side note, it looks like we treat `apiVersion` differently. If I run through the same exercise with `apiVersion`, I do get a warning when using a parameter / variable for the value. We have a current bug for this - https://github.com/microsoft/vscode-azurearmtools/issues/485

![image](https://user-images.githubusercontent.com/7844635/79365888-1ce94a80-7f00-11ea-93cb-4d715876cca3.png)

**Comment by nilshedstrom** [comment]
Should we require the user to select the text that should be extracted into a variable or parameter?
Or should we propose different possibilities based on the current context? 
This is how it looks when you extract a variable in Visual Studio (using Resharper) without selecting anything. The cursor was at in the Message string when extract variable was invoked.
![image](https://user-images.githubusercontent.com/785949/79633972-080ef000-8168-11ea-8c82-6d96322110ae.png)

I would like to work on this issue but since it is dependent on #589 and a little bit on #621 I would like to wait untill thoose are merged into master.

**Comment by nilshedstrom** [comment]
I started implementing a fix for this issue in https://github.com/nilshedstrom/vscode-azurearmtools/tree/extractVariableOrParameter

**Comment by neilpeterson** [comment]
@nilshedstrom we will review soon, thanks.

**Comment by StephenWeatherford** [comment]
@nilshedstrom Just started playing with this, so far looking great!  Some quick observations for behavior:

1) Bug:
```json
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "resources": [
        {
            "apiVersion": "2015-01-01",
            "name": "fetchSecret",
            "type": "Microsoft.Resources/deployments"
        }
    ],
    "outputs": {
    }
}
```
Extract "fetchSecret" and you get this: 
```
    "outputs": ,
    "parameters": {
        "p1": {
            "type": "string",
            "defaultValue": "fetchSecret"
        }
    }{
    }
```
2) After extracting, Undo should ideally undo both of the edits at once.  There's a flag in the texteditor edit command that's supposed to allow this, although I had trouble with it before.
3) You'll need to support nested templates with outer or inner scope and also user functions.  You should wait until https://github.com/microsoft/vscode-azurearmtools/pull/826 is merged.  TemplateScope.rootObject is part of what you will need.

 For example:
```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "resources": [
        {
            "apiVersion": "2015-01-01",
            "name": "fetchSecret",
            "type": "Microsoft.Resources/deployments",
            "properties": {
                "expressionEvaluationOptions": {
                    "scope": "inner"
                },
                "mode": "Incremental",
                "template": {
                    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
                    "contentVersion": "1.0.0.0",
                    "parameters": {
                        // <<<<<<<<<<<<< SHOULD END UP HERE (if it's not inner-scoped, it should end up in the first non-outer-scoped parent or top level)
                    },
                    "resources": [
                    ],
                    "outputs": {
                        "secretValueOutput": {
                            "type": "secureString",
                            "value": "abc"  // <<<<<<<<<<<<<< EXTRACT THIS
                        }
                    }
                }
            }
        }
    ]
}
```
4) I think we should show the extract action whenever the user is inside a non-expression string as well, i.e. as if the entire string contents were selected (even if no text is selected).  This will be a common scenario, and will also make it much more discoverable.
5) I think these should also show up in the editor context menu


P2 to consider (@neilpeterson  Thoughts? - maybe we just put them in the backlog)
1) Handle substrings inside of an existing expression
2) If there's a param or variable with the same (default) text, offer to change it to the existing parameter/variable
3) Automatically find matching strings for #2

**Comment by neilpeterson** [comment]
@nilshedstrom we are starting on 0.13.0 today. Do you think you will be able to address this in the upcoming weeks?

**Comment by nilshedstrom** [comment]
Yes, I should have some time to work on this soon. 

Do you have some kind of deadline?

**Comment by StephenWeatherford** [comment]
Not specifically, although if you want it in this next release, we'd need it within the next 2-3 weeks.  THanks!

**Comment by StephenWeatherford** [comment]
@nilshedstrom FYI, if you still have time, we need it within a week for testing.  Thx.

**Comment by nilshedstrom** [comment]
> 2\. You'll need to support nested templates with outer or inner scope and also user functions

@StephenWeatherford 
Can you explain what you mean by support for user functions.
At the moment you can only extrct variables or parameters in the resource section.
As I understand the schema you cannot have variables inside your user function and extracting parameters inside a user function does not sound like a top priority to me.

**Comment by StephenWeatherford** [comment]
> > 2. You'll need to support nested templates with outer or inner scope and also user functions
> 
> @StephenWeatherford
> Can you explain what you mean by support for user functions.
> At the moment you can only extrct variables or parameters in the resource section.
> As I understand the schema you cannot have variables inside your user function and extracting parameters inside a user function does not sound like a top priority to me.

I agree it's not a top priority, but we should either handle it or disallow it.  Try extracting the 'suffix' string from this user function: https://github.com/StephenWeatherford/template-examples/blob/67fce55c328d9fa45625a32bea29a8485285278b/nested-userfuncs.json#L65

Notice that because it's inside a resource (the top-level deployments resource), we think we can extract it but we extract to the top level.  Either we fix it so we don't handle this scenario (and correctly recognize it's not in a supported scope here), or we allow outside resources and fix it to extract params only to the correct scope, which shouldn't be too hard either, but given the lower priority either is fine.  Or consider it a P2 bug to fix later.

**Comment by nilshedstrom** [comment]
I fixed the bug. Now parameters/variables can only be extracted if they are in the resources section of the current scope.

**Comment by StephenWeatherford** [comment]
Fixed by @nilshedstrom in #https://github.com/microsoft/vscode-azurearmtools/pull/976
